src/main/java/org/jitsi/videobridge/xmpp/MediaStreamTrackFactory.java

```diff
@@ -383,157 +106,159 @@ private static void removeReferences(TrackSsrcs trackSsrcs, List<SourcePacketExt
+    public MediaStreamTrackFactory(
+            MediaStreamTrackReceiver trackReceiver,
+            List<SourcePacketExtension> sources,
+            List<SourceGroupPacketExtension> sourceGroups)
     {
-        List<TrackSsrcs> trackSsrcsList = getTrackSsrcs(sources, sourceGroups);
-        List<MediaStreamTrackDesc> tracks = new ArrayList<>();
-        trackSsrcsList.forEach(trackSsrcs -> {
-            // As of now, we only ever have 1 spatial layer per stream
-            int numSpatialLayersPerStream = 1;
-            int numTemporalLayersPerStream = 1;
-            if (trackSsrcs.size() > 1 && ENABLE_SVC)
-            {
-                numTemporalLayersPerStream = VP8_SIMULCAST_TEMPORAL_LAYERS;
-            }
-            Map<Long, SecondarySsrcs> secondarySsrcs = getAllSecondarySsrcs(trackSsrcs,
-                sourceGroups);
-            MediaStreamTrackDesc track = createTrack(mediaStreamTrackReceiver,
-                trackSsrcs, numSpatialLayersPerStream,
-                numTemporalLayersPerStream, secondarySsrcs);
-            tracks.add(track);
-        });
-
-        return tracks.toArray(new MediaStreamTrackDesc[tracks.size()]);
+        this.trackReceiver = trackReceiver;
+        this.sources = sources;
+        this.sourceGroups = sourceGroups;
     }
 
     /**
-     * Calculates the subjective quality index of an RTP flow specified by its
-     * stream index (simulcast), spatial index (SVC) and temporal index (SVC).
-     *
-     * @param streamIdx the stream index.
-     * @param spatialIdx the spatial layer index.
-     * @param temporalIdx the temporal layer index.
+     * Creates {@link MediaStreamTrackDesc}s from signaling params
      *
-     * @return the subjective quality index of the flow specified in the
-     * arguments.
+     * @return the {@link MediaStreamTrackReceiver} that are described in the
+     * jingle sources and source groups.
      */
-    private static int qid(int streamIdx, int spatialIdx, int temporalIdx,
-                           int spatialLen, int temporalLen)
+    public MediaStreamTrackDesc[] createMediaStreamTracks()
     {
-        return streamIdx * spatialLen * temporalLen
-            + spatialIdx * temporalLen + temporalIdx;
-    }
+        tracksMap.clear();
+
+	if (sourceGroups != null && !sourceGroups.isEmpty())
+	{
+            // Make a copy of the groups because we delete any SIM groups we
+            // find.
+            List<SourceGroupPacketExtension> groupsCopy
+                = new ArrayList<>(sourceGroups);
+
+            // We first process the special SIM group. The SIM group represents
+            // a single track (or video source) it has 3 spatial encodings and
+            // each spatial encoding has 3 temporal encodings. The temporal
+            // sub-encodings share the SSRCs of the parent spatial encoding. 
+	    Iterator<SourceGroupPacketExtension> it = groupsCopy.iterator();
+
+	    while (it.hasNext())
```

Comment:
I find iterators hard to follow. How about checking the semantics in addVP8SimulcastTrack and returning a boolean? Then this iterator block will become just this:
```
groupsCopy.removeIf(this::addVP8SimulcastTrack);
```

Reply:
i agree that it'd be nice to make the loop more readable, but not sure i like calling `addVP8SimulcastTrack` for every track and it just returning true for the ones that actually were simulcast tracks.  i think it'd be clearer to do these sorts of things more procedurally, like:

```
  // sim group processing
  simGroups = getSimGroups(sourceGroups)
  (do the work with the sim groups)
  sourceGroups.remove(simGroups)
  
  // rtx/fec processing
  ...
```